lib/restforce/concerns/sobject_collection_api.rb

```diff
@@ -0,0 +1,149 @@
+
+      def collection_create(opts = {})
+        all_or_none = opts.delete(:all_or_none) || false
+        builder = RecordsBuilder.new
+        yield(builder)
+        return builder.records if opts[:debug]
+
+        if builder.records.empty?
+          raise ArgumentError, "There are no records to be created"
+        end
+
+        results = api_post('composite/sobjects',
+                           {
+                             allOrNone: all_or_none,
+                             records: builder.records
+                           }).body
+        CollectionResponse.new(results, all_or_none: all_or_none).response
+      end
+
+      def collection_create!(opts = {}, &block)
+        collection_create(opts.merge(all_or_none: true), &block)
+      end
+
+      def collection_update(opts = {})
+        all_or_none = opts.delete(:all_or_none) || false
+        builder = RecordsBuilder.new
+        yield(builder)
+        return builder.records if opts[:debug]
+
+        if builder.records.empty?
+          raise ArgumentError, "There are no records to be created"
+        end
+
+        results = api_patch('composite/sobjects',
+                            {
+                              allOrNone: all_or_none,
+                              records: builder.records
+                            }).body
+        CollectionResponse.new(results, all_or_none: all_or_none).response
+      end
+
+      def collection_update!(opts = {}, &block)
+        collection_update(opts.merge(all_or_none: true), &block)
+      end
+
+      def collection_upsert(sobject_type, field_name, opts = {})
+        all_or_none = opts.delete(:all_or_none) || false
+        builder = RecordsBuilder.new(field_name.to_sym)
+        yield(builder)
+        return builder.records if opts[:debug]
+
+        if builder.records.empty?
+          raise ArgumentError, "There are no records to be created"
+        end
+
+        results = api_patch("composite/sobjects/#{sobject_type}/#{field_name}",
+                            {
+                              allOrNone: all_or_none,
+                              records: builder.records
+                            }).body
+        CollectionResponse.new(results, all_or_none: all_or_none).response
+      end
+
+      def collection_upsert!(sobject_type, field_name, opts = {}, &block)
+        collection_upsert(sobject_type, field_name, opts.merge(all_or_none: true), &block)
+      end
+
+      class CollectionResponse
+        attr_accessor :results, :all_or_none
+
+        def initialize(results, all_or_none: false)
+          @results = results
+          @all_or_none = all_or_none
+        end
+
+        def response
+          has_errors = results.any? { |result| !result.success }
+          if all_or_none && has_errors
+            last_error_index = results.rindex do |result|
+              !result.success && !result.errors.empty? &&
+                result.errors.last.statusCode != 'ALL_OR_NONE_OPERATION_ROLLED_BACK '
```

Comment:
Is the extra space here intentional?
```suggestion

                result.errors.last.statusCode != 'ALL_OR_NONE_OPERATION_ROLLED_BACK'
```